pkg/scheduler/queue/tree_queue_algo_querier_worker_queue_priority_test.go

```diff
@@ -0,0 +1,345 @@
+		{dequeue, QueuePath{"child-3"}, "obj-3"},
+
+		// enqueue for child-1 again to verify it is added back to rotation
+		{enqueue, QueuePath{"child-1"}, "obj-7"},
+
+		// child-3 is still next; child-1 was added back to rotation
+		{dequeue, QueuePath{"child-3"}, "obj-4"},
+		// child-3 is now empty and removed from rotation; only child-1 remains
+		{dequeue, QueuePath{"child-1"}, "obj-7"},
+		// nothing left to dequeue
+		{dequeue, QueuePath{}, nil},
+	}
+
+	querierWorkerPrioritizationQueueAlgo := NewQuerierWorkerQueuePriorityAlgo()
+
+	tree, err := NewTree(querierWorkerPrioritizationQueueAlgo, &roundRobinState{})
+	require.NoError(t, err)
+
+	for _, operation := range operationOrder {
+		if operation.kind == enqueue {
+			err = tree.EnqueueBackByPath(operation.path, operation.obj)
+			require.NoError(t, err)
+		}
+		if operation.kind == dequeue {
+			path, obj := tree.Dequeue()
+			require.Equal(t, operation.path, path)
+			require.Equal(t, operation.obj, obj)
+		}
+	}
+}
+
+// Test for the expected behavior of the queue algorithm when there are multiple workers
+// and when the querier-worker queue priority algorithm is used only at the highest layer of the tree.
+// When at the highest layer of the tree, the global node count for any existing node is always 1,
+// so the queue algorithm will always delete the node from the global node order rotation after it is emptied.
+func TestQuerierWorkerQueuePriority_StartPositionByWorker(t *testing.T) {
+	querierWorkerPrioritizationQueueAlgo := NewQuerierWorkerQueuePriorityAlgo()
+
+	tree, err := NewTree(querierWorkerPrioritizationQueueAlgo, &roundRobinState{})
+	require.NoError(t, err)
+
+	// enqueue 3 objects each to 3 different children;
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{ingesterQueueDimension}, "obj-1"))                // ingester node created
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{storeGatewayQueueDimension}, "obj-2"))            // store-gateway node created
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{ingesterAndStoreGatewayQueueDimension}, "obj-3")) // ingester-and-store-gateway node created
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{ingesterQueueDimension}, "obj-4"))
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{storeGatewayQueueDimension}, "obj-5"))
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{ingesterAndStoreGatewayQueueDimension}, "obj-6"))
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{ingesterQueueDimension}, "obj-7"))
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{storeGatewayQueueDimension}, "obj-8"))
+	require.NoError(t, tree.EnqueueBackByPath(QueuePath{ingesterAndStoreGatewayQueueDimension}, "obj-9"))
+
+	// node order was set by initial enqueue order;
+	// order will remain until a node is deleted for being empty or a new node is added by an enqueue
+	expectedInitialNodeOrder := []string{ingesterQueueDimension, storeGatewayQueueDimension, ingesterAndStoreGatewayQueueDimension}
```

Comment:
Do you need this check? You can also essentially check order via dequeuing objects (you expect the first thing dequeued by querier-worker 0 to be from `ingesterQueueDimension`, etc), and it involves less peeping of internals.

Reply:
So technically yes, *anything* could be happening to the node order behind the scenes and as long as the workerID->deueueNode relation holds from outside observation, then it's working.

However because the dequeue order depends on which workerID is dequeuing, I want both checks - that the order has not been changed when it should not be changed, and that the queue selection by workerID is correct.

I generally take the opinion that these lowest-level unit tests are also supposed to serve as unambiguous documentation of intent for how something is intended to work, even if that behavior is not necessarily visible when observing the system from a wider angle. It's for the next engineer who works on it to understand the intended inner machinations a bit better.

This is the lowest-level most contained direct unit test of this class, so I believe this is the right place to peek internals (when it matters). I would not want to assert on this in `queue_test.go` or `multi_queuing_algorithm_tree_queue_test.go` or the benchmarks, etc.